xds/internal/xdsclient/tlscreds/bundle.go

```diff
@@ -0,0 +1,138 @@
+/*
+ *
+ * Copyright 2023 gRPC authors.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ *
+ */
+
+// Package tlscreds implements mTLS Credentials in xDS Bootstrap File.
+// See gRFC A65: github.com/grpc/proposal/blob/master/A65-xds-mtls-creds-in-bootstrap.md.
+package tlscreds
+
+import (
+	"context"
+	"crypto/tls"
+	"encoding/json"
+	"fmt"
+	"net"
+
+	"google.golang.org/grpc/credentials"
+	"google.golang.org/grpc/credentials/tls/certprovider"
+	"google.golang.org/grpc/credentials/tls/certprovider/pemfile"
+)
+
+// bundle is an implementation of credentials.Bundle which implements mTLS
+// Credentials in xDS Bootstrap File.
+type bundle struct {
+	transportCredentials credentials.TransportCredentials
+}
+
+// NewBundle returns a credentials.Bundle which implements mTLS Credentials in xDS
+// Bootstrap File. It delegates certificate loading to a file_watcher provider
+// if either client certificates or server root CA is specified.
+// See gRFC A65: github.com/grpc/proposal/blob/master/A65-xds-mtls-creds-in-bootstrap.md
+func NewBundle(jd json.RawMessage) (credentials.Bundle, error) {
+	cfg := &struct {
+		CertificateFile   string `json:"certificate_file"`
+		CACertificateFile string `json:"ca_certificate_file"`
+		PrivateKeyFile    string `json:"private_key_file"`
+	}{}
+
+	if err := json.Unmarshal(jd, cfg); err != nil {
+		return nil, err
```

Comment:
Should we add some context to the error here. The error returned by the `json` package might simply look like:
`Unmarshal failed:  unexpected end of JSON input`

So, maybe we can do:
`return nil, fmt.Errorf("tlscreds: failed to unmarshal config: %v", err)`

Reply:
Good call, although I stripped tlscreds from what you suggested since the error message currently looks like this:

```
failed to exit idle mode: xds: failed to create xds-client: xds: failed to read bootstrap file: xds: json.Unmarshal(data) for field "xds_servers" failed during bootstrap: failed to unmarshal JSON to []*ServerConfig: failed to build credentials bundle from bootstrap for "tls": <this message>
```

So for users the package name (tlscreds) is irrelevant here.

So now with the following invalid config:

```

 the error message looks like this:

```
failed: failed to exit idle mode: xds: failed to create xds-client: xds: failed to read bootstrap file: xds: json.Unmarshal(data) for field "xds_servers" failed during bootstrap: failed to unmarshal JSON to []*ServerConfig: failed to build credentials bundle from bootstrap for "tls": failed to unmarshal config: json: cannot unmarshal array into Go value of type struct { CertificateFile string "json:\"certificate_file\""; CACertificateFile string "json:\"ca_certificate_file\""; PrivateKeyFile string "json:\"private_key_file\"" }
```

Reply:
Good call to add some context. I took your suggestion, although I didn't include `tlscreds` because it seems irrelevant to users, even though I saw that prefixing error messages with the package name is a fairly common thing to do. Here there's already the bundle name in the error message that users receive.

With an invalid config due to json not fitting in the struct, such as:

```
{
  "xds_servers": [{
    "server_uri": "localhost:4040",
    "channel_creds": [{
      "type": "tls",
      "config": []
    }]
  }]
}

``` 

we get the following error message:

```
failed: failed to exit idle mode: xds: failed to create xds-client: xds: failed to read bootstrap file: xds: json.Unmarshal(data) for field "xds_servers" failed during bootstrap: failed to unmarshal JSON to []*ServerConfig: failed to build credentials bundle from bootstrap for "tls": failed to unmarshal config: json: cannot unmarshal array into Go value of type struct { CertificateFile string "json:\"certificate_file\""; CACertificateFile string "json:\"ca_certificate_file\""; PrivateKeyFile string "json:\"private_key_file\"" }
```